docs/proposal/AI-gateway.md

```diff
@@ -0,0 +1,206 @@
+
+So we finally chose Envoy's [External Processing Filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter). It connects to an external service, called an "external processor", to the filter chain. The processing service itself implements a gRPC interface that allows it to respond to events in the lifecycle of an HTTP request/response by examining and modifying the headers, body, and trailers of each message, or by returning a brand-new response.
+
+In this way, our AI plugin can be built as an external independent service and can be deployed on demand when AI traffic needs to be processed. At the same time, we can use any dev language, not necessarily C++. Of course, we use Golang. Also it is isolated from the complexity of Envoy to greatest extent, and only needs to process traffic requests/responses from gRPC connections. This fully ensures that development is simple, flexible and friendly.
+
+#### AI Plugin CRD
+
+We need to rely on CRD to describe the config related to AI traffic management. The config consists of the following two parts:
+
+1. Associated Service: The AI model is ultimately deployed in a service. We hope that only the service configured with the corresponding AI policy will forward traffic to the AI plugin and all AI services in the cluster can share the same AI plugin deployment without having to deploy one for each Service. In addition, if a service configured with AI policy, traffic can be directly forwarded to the AI plugin. Whether it's through automatically generating resource objects such as `EnvoyExtensionPolicy` or modifying XDS. It's believed that it will bring a better user experience.
+
+2. Specific AI traffic management policy: such as rate limit, cache, fallback, etc. Allow users to configure any number and type of policies.
+
+API:
+
+```go
+// AIExtension is the Schema for the aiextensions API.
+type AIExtension struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   AIExtensionSpec   `json:"spec,omitempty"`
+	Status AIExtensionStatus `json:"status,omitempty"`
+}
+
+// AIExtensionSpec defines the desired state of AIExtension.
+type AIExtensionSpec struct {
+	Options Options `json:"options"`
+}
+
+type Options struct {
+	RateLimits []RateLimit `json:"rateLimits,omitempty"`
+  // Add more polices as needed.
+}
+
+// +kubebuilder:validation:Enum=second;minute;hour;day;month
+type RateLimitUnit string
+
+const (
+	Second RateLimitUnit = "second"
+	Minute RateLimitUnit = "minute"
+	Hour   RateLimitUnit = "hour"
+	Day    RateLimitUnit = "day"
+	Month  RateLimitUnit = "month"
+)
+
+// RateLimit Policy for AI traffic.
+type RateLimit struct {
+	RequestsPerUnit uint32        `json:"requestsPerUnit"`
+	Unit            RateLimitUnit `json:"unit"`
+	Model           string        `json:"model"`

```

Comment:
how do you express token based rate limit

Reply:
Will rename to `tokensPerUnit`.